accessEntity/data.go

```diff
@@ -0,0 +1,751 @@
+			fallthrough
+		case "":
+			apiProducts, err = d.getApiProductsByAppName(priVal, "", "", "", org)
+		}
+		if err != nil {
+			return
+		}
+	} else if priKey == IdentifierConsumerKey {
+		apiProducts, err = d.getApiProductsByConsumerKey(priVal, org)
+		if err != nil {
+			return
+		}
+	}
+
+	if secKey == IdentifierApiResource {
+		apiProducts = filterApiProductsByResource(apiProducts, secVal)
+	}
+	return
+}
+
+func (d *DbManager) GetApps(org, priKey, priVal, secKey, secVal string) (apps []common.App, err error) {
+	switch priKey {
+	case IdentifierAppId:
+		return d.getAppByAppId(priVal, org)
+	case IdentifierAppName:
+		switch secKey {
+		case IdentifierDeveloperEmail:
+			return d.getAppByAppName(priVal, secVal, "", "", org)
+		case IdentifierDeveloperId:
+			return d.getAppByAppName(priVal, "", secVal, "", org)
+		case IdentifierCompanyName:
+			return d.getAppByAppName(priVal, "", "", secVal, org)
+		case "":
+			return d.getAppByAppName(priVal, "", "", "", org)
+		}
+	case IdentifierConsumerKey:
+		return d.getAppByConsumerKey(priVal, org)
+	}
+	return
+}
+
+func (d *DbManager) GetCompanies(org, priKey, priVal, secKey, secVal string) (companies []common.Company, err error) {
+	switch priKey {
+	case IdentifierAppId:
+		return d.getCompanyByAppId(priVal, org)
+	case IdentifierCompanyName:
+		return d.getCompanyByName(priVal, org)
+	case IdentifierConsumerKey:
+		return d.getCompanyByConsumerKey(priVal, org)
+	}
+	return
+}
+
+func (d *DbManager) GetCompanyDevelopers(org, priKey, priVal, secKey, secVal string) (companyDevelopers []common.CompanyDeveloper, err error) {
+	if priKey == IdentifierCompanyName {
+		return d.getCompanyDeveloperByComName(priVal, org)
+	}
+	return
+}
+
+func (d *DbManager) GetAppCredentials(org, priKey, priVal, secKey, secVal string) (appCredentials []common.AppCredential, err error) {
+
+	switch priKey {
+	case IdentifierConsumerKey:
+		return d.getAppCredentialByConsumerKey(priVal, org)
+	case IdentifierAppId:
+		return d.getAppCredentialByAppId(priVal, org)
+	}
+	return
+}
+
+func (d *DbManager) GetDevelopers(org, priKey, priVal, secKey, secVal string) (developers []common.Developer, err error) {
+	switch priKey {
+	case IdentifierAppId:
+		return d.getDeveloperByAppId(priVal, org)
+	case IdentifierDeveloperEmail:
+		return d.getDeveloperByEmail(priVal, org)
+	case IdentifierConsumerKey:
+		return d.getDeveloperByConsumerKey(priVal, org)
+	case IdentifierDeveloperId:
+		return d.getDeveloperById(priVal, org)
+	}
+	return
+}
+
+func (d *DbManager) getApiProductsByName(apiProdName string, org string) (apiProducts []common.ApiProduct, err error) {
+	err = d.GetDb().QueryStructs(&apiProducts,
+		sql_select_api_product+
+			`WHERE ap.name = ? AND ap.tenant_id IN `+
+			sql_select_tenant_org,
+		apiProdName,
+		org,
+	)
+	return
+}
+
+func (d *DbManager) getApiProductsByAppId(appId string, org string) (apiProducts []common.ApiProduct, err error) {
+	cols := []string{"*"}
+	query := selectApiProductsById(
```

Comment:
Such query construction hard to read/follow/maintain...
please re-approach on this one.

Reply:
Can you review and re-factor this?

Reply:
@chebrolus 
My previous comment was folded due to code change, which was on another thread to illustrate this query construction design. So I paste it here:

The motivation was to reuse code, and increase extensibility.
In fact, each access entity request has different logics according to identifiers. Sometimes there're 2 or 3 table joins happenning.
A straightforward way is to hard-code all nested long SQL queries for each entity-primaryIdentifier-secondaryIdentifier combination.
However, if there's any change in db schema, all affected SQL queries have to be changed. In case new identifiers are added, this will also be a non-extensible manual process.
We discussed about the http://graphql.org/ during API design, but that would be too much work for MP. So we decided to use normal http GET for accessEntity. However, ideally in the future this will be the only portal to access KMS data, and there will be a set of generic apis to do that. So I designed it in this way to enable it generate SQL statements at runtime.
Either we adopt graphql in future or not, the nested sql construction in a stack manner is should be a good start. The generic logics & extensible design are on our roadmap to a final generic plugin that exposes KMS data.